source/client/world/ClientPlayer.cpp

```diff
@@ -72,54 +72,89 @@ void ClientPlayer::updateCamera() {
+			if (m_velocity.z < 0.0f) {
+				if (m_isGliding) m_isGliding = false;
+			}
+		}
 	}
 
-	if(gk::GamePad::isKeyPressed(GameKey::Forward))    move(0.0f);
-	else if(gk::GamePad::isKeyPressed(GameKey::Back))  move(180.0f);
+	if(gk::GamePad::isKeyPressed(GameKey::Forward))    move(world, 0.0f);
+	else if(gk::GamePad::isKeyPressed(GameKey::Back))  move(world, 180.0f);
 
-	if(gk::GamePad::isKeyPressed(GameKey::Left))       move(90.0f);
-	else if(gk::GamePad::isKeyPressed(GameKey::Right)) move(-90.0f);
+	if(gk::GamePad::isKeyPressed(GameKey::Left))       move(world, 90.0f);
+	else if(gk::GamePad::isKeyPressed(GameKey::Right)) move(world, -90.0f);
 
-	if (gk::GamePad::isKeyPressed(GameKey::Left)  && gk::GamePad::isKeyPressed(GameKey::Forward)) move(45.0f);
-	if (gk::GamePad::isKeyPressed(GameKey::Right) && gk::GamePad::isKeyPressed(GameKey::Forward)) move(-45.0f);
-	if (gk::GamePad::isKeyPressed(GameKey::Left)  && gk::GamePad::isKeyPressed(GameKey::Back))    move(135.0f);
-	if (gk::GamePad::isKeyPressed(GameKey::Right) && gk::GamePad::isKeyPressed(GameKey::Back))    move(-135.0f);
+	if (gk::GamePad::isKeyPressed(GameKey::Left)  && gk::GamePad::isKeyPressed(GameKey::Forward)) move(world, 45.0f);
+	if (gk::GamePad::isKeyPressed(GameKey::Right) && gk::GamePad::isKeyPressed(GameKey::Forward)) move(world, -45.0f);
+	if (gk::GamePad::isKeyPressed(GameKey::Left)  && gk::GamePad::isKeyPressed(GameKey::Back))    move(world, 135.0f);
+	if (gk::GamePad::isKeyPressed(GameKey::Right) && gk::GamePad::isKeyPressed(GameKey::Back))    move(world, -135.0f);
 
-	if (gk::GamePad::isKeyPressed(GameKey::Sprint)) {
-		m_velocity.x *= 1.5f;
-		m_velocity.y *= 1.5f;
+	if (chunk && chunk->isInitialized()) {
+		if (gk::GamePad::isKeyPressed(GameKey::Sprint)) {
+			m_velocity.x *= chunk->dimension().physics().horizontalSprintMod;
+			m_velocity.y *= chunk->dimension().physics().verticalSprintMod;
+		}
+
+		if(gk::GamePad::isKeyPressed(GameKey::Fly)) {
+			m_velocity.z = chunk->dimension().physics().flySpeed;
+		}
+
+		if(gk::GamePad::isKeyPressed(GameKey::Sneak)) {
+			if (m_velocity.z == 0 || chunk->dimension().physics().isSneakAlwaysMod) {
+				m_velocity.x *= chunk->dimension().physics().sneakHorizontalMod;
+				m_velocity.y *= chunk->dimension().physics().sneakHorizontalMod;
+			}
+
+			m_velocity.z = -chunk->dimension().physics().sneakVerticalSpeed;
+		}
 	}
 }
 
 void ClientPlayer::updatePosition(const ClientWorld &world) {
 	ClientChunk *chunk = (ClientChunk *)world.getChunkAtBlockPos(m_x, m_y, m_z);
 	if (chunk && chunk->isInitialized()) {
 		if (!Config::isFlyModeEnabled) {
-			m_velocity.z -= chunk->dimension().gravity() * 0.001f;
+			m_velocity.z -= chunk->dimension().physics().gravity * 0.001f;
 
 			m_isJumping = true;
 
-			if (m_velocity.z < -m_jumpSpeed) // Limit max vertical speed to jump speed
-				m_velocity.z = -m_jumpSpeed;
+			if (m_isGliding) {
+				if (m_velocity.z < -chunk->dimension().physics().glideGravity) // Limit max vertical speed to jump speed
```

Comment:
`to jump speed` => `to glide gravity`?

Reply:
The idea is that if you're gliding your max fall speed is glideGravity, otherwise it's jumpSpeed. Basically, I added a system where holding space slows your fall speed.

Reply:
Yeah that's not the issue, look at the comment, it's wrong

Reply:
What do you mean? Is negative velocity an upwards vector?

Reply:
Exactly. :')